Grafana/icons: Add icons package

Clarity-89 opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

An alternative to https://github.com/grafana/icons/pull/12, where the new icons package is located inside the Grafana repository. 

**How it works**

1. SVGs for the icons are uploaded to the `packages/grafana-icons/svg` folder.
2. Running `yarn generate` converts the SVGs into TypeScript React components inside the `packages/grafana-icons/src/icons-generated` folder. Each component is wrapped in an `IconBase` component for easier customisation of the icons. A custom template in `packages/grafana-icons/templates/icon.js` is used to generate each component.
3. `yarn build` compiles the generated code and creates esm + cjs bundles in the `packages/grafana-icons/dist` folder. This is mostly a copy/paste of the same logic from `@grafana` packages.
4. The CI pipeline is out of scope of this PR to keep it smaller.

Fixes: https://github.com/grafana/grafana/issues/81682.


**Questions/issues**
- Do all the icons need to be wrapped in `forwardRef`? We currently do it for the `Icon` component, but I wonder if it's used? It will add quite a bit of complexity to the icon template. 
- Do we need both cjs and esm builds? 

**Workflow diagram:**

```mermaid
flowchart TD
    A[An SVG file for an icon is uploaded <br> to the `svg` folder.]

    B[Run 'yarn generate' script <br> to use the components locally.]
    C[New PR created and merged]
    D['generate' script is run<br> to generate the components]
    E[TS files are transpiled to JS <br> and bundled into an ESM module ]
    G[New version of the package is published]

    A --> |Local dev| B
    B --> C
    A --> C
    C --> |CI|D
    D --> |CI|E
    E --> |CI|G
```

Clarity-89 committed:
Add separate icons package

Clarity-89 committed:
Update package.json

Clarity-89 committed:
Codeformat

Clarity-89 committed:
Use the new icon

Clarity-89 committed:
Merge branch 'main' into add-icons-package

Clarity-89 committed:
Sanitize SVGs

Clarity-89 committed:
Sync deps

Clarity-89 committed:
Add publish script

Clarity-89 committed:
Add pre-commit hook

Clarity-89 committed:
Test pre-commit

Clarity-89 committed:
Rename

Clarity-89 committed:
Comment out local dev configs

Clarity-89 committed:
Add generate action

Clarity-89 committed:
Update readme

Clarity-89 committed:
Update script

Clarity-89 committed:
Add license

Clarity-89 committed:
Rename icons to icons-generated

Clarity-89 commented on .github/workflows/pr-generate-icons.yml:
```diff
@@ -0,0 +1,40 @@
+name: Generate icon components
```
Haven't tested this and the publish pipelines yet, but this is the general idea of how they'd work.

Clarity-89 commented on package.json:
```diff
@@ -424,6 +425,7 @@
   "workspaces": {
     "packages": [
       "packages/*",
+      "packages/!(grafana-icons)/**",
```
This will exclude the icons package from being versioned and published by lerna.

Clarity-89 replied:
Looks like this requires upgrading to lerna v8 to work.

Clarity-89 commented on packages/grafana-icons/package.json:
```diff
@@ -0,0 +1,60 @@
+{
+  "name": "@grafana/icons",
```
@staton-hyse11 this is the name under which the package will be published. Which ideas for the name did you have in mind?

staton-hyse11 replied:
@Clarity-89 "@grafana/saga-icons" makes sense to me here. it keeps the pattern of "@grafana" but notes these are part of the design system.

Clarity-89 commented on public/app/core/components/Breadcrumbs/BreadcrumbItem.tsx:
```diff
@@ -44,7 +45,7 @@ export function BreadcrumbItem({ href, isCurrent, text, index, flexGrow }: Props
           </a>
           <div className={styles.separator} aria-hidden={true}>
-            <Icon name="angle-right" />
+            <AngleRight />
```
This is the usage example of the new icons.

Clarity-89 reviewed (commented):
@joshhunt @academo I've also included your feedback from the standalone repo PR into this PR.

Clarity-89 committed:
Codeformat

Clarity-89 committed:
Merge branch 'main' into add-icons-package

Clarity-89 committed:
Update lerna

ashharrison90 reviewed (commented):
@Clarity-89 lerna upgrade probably should be done separately as it caused an issue with package publishing last time, see https://github.com/grafana/grafana/pull/82254

believe @jackw is also looking at it as part of https://github.com/grafana/grafana/pull/81873

jackw commented on package.json:
```diff
@@ -181,7 +181,7 @@
     "jest-junit": "16.0.0",
     "jest-matcher-utils": "29.7.0",
-    "lerna": "7.4.1",
+    "lerna": "^8.1.2",
```
FYI - I believe this update will break package publishing. I also need to upgrade it to land [this PR](https://github.com/grafana/grafana/pull/81873). Maybe best to be tackled in a separate PR first? 🤔

Clarity-89 commented:
> @Clarity-89 lerna upgrade probably should be done separately as it caused an issue with package publishing last time, see #82254
> 
> believe @jackw is also looking at it as part of #81873

Yep, I'll wait until we have that working, the CI wont pass here until that upgrade is done (or we move the icon to a separate dir). I was mainly  testing if CI will pass after the upgrade.

Clarity-89 committed:
Remove fill attribute from SVG path elements

Clarity-89 committed:
Revert lerna upgrade

Clarity-89 committed:
Update codeowners

Clarity-89 committed:
Merge branch 'main' into add-icons-package

Clarity-89 committed:
Add lint scripts

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	package.json
#	yarn.lock

Clarity-89 committed:
Exclude more attributes

jackw commented on packages/grafana-icons/rollup.config.ts:
```diff
@@ -0,0 +1,35 @@
+import resolve from '@rollup/plugin-node-resolve';
```
As it is right now it looks like we have a pipeline of: SVG -> TS -> JS + TS definitions.

Do we need rollup and typescript for this package? Could we ditch them both in favour of making this package an ESM only package, and set svgr to output to JS? I don't think there's any of our tooling that can't work with this setup and ESM for Icons will give much better tree-shaking potential than the CJS output. There is the rest spread in `IconBase` but not sure how problematic that would be if we're not transpiling.

As for the types the TS definitions are quite basic with a lot of repetition. Could we add a simple script that generates this too?

```
import React, { SVGProps } from 'react';

type IconSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl' | 'xxxl';
interface IconProps extends Omit<SVGProps<SVGSVGElement>, 'onLoad' | 'onError' | 'ref'> {
  size?: IconSize;
  title?: string;
  color?: string;
}

declare const SvgAngleRight: (props: IconProps) => React.JSX.Element;
declare const SvgApps: (props: IconProps) => React.JSX.Element;
declare const SvgBook: (props: IconProps) => React.JSX.Element;
declare const SvgCancel: (props: IconProps) => React.JSX.Element;
export { SvgAngleRight as AngleRight, SvgApps as Apps, SvgBook as Book, SvgCancel as Cancel, type IconProps };

```
I might be missing something but I think in this way the package could be stripped down to only the svgs as src and IconBase.js in the root with the icon components and the index.d.ts generated from a single build command.

Clarity-89 replied:
> Do we need rollup and typescript for this package?

I've been wondering the same and the reasoning for adding it is mainly to use types in IconBase. Also it helps to validate that generated components are properly typed since we've customised the component template. 


> As for the types the TS definitions are quite basic with a lot of repetition. Could we add a simple script that generates this too?

Do you mean to generate index.d.ts that provides types for the generated Icons and IconBase? I can look into that, should be pretty straightforward and simplify the package and CI a lot, altho it does mean losing type safety here. I guess we can always add it back if the component logic becomes too complex 🤔 We can also look into using JSDoc comments to add some type safety without the build step.

jackw replied:
>altho it does mean losing type safety here

Would we lose type safety? Won't the package.json reference the index.d.ts file which would be pretty much the same as the currently built index.d.ts output?

Clarity-89 replied:
Will it apply when working on the IconBase and other files in the icons package? If yes, then I guess we're all good.

Clarity-89 replied:
> I might be missing something but I think in this way the package could be stripped down to only the svgs as src and IconBase.js in the root with the icon components and the index.d.ts generated from a single build command.

We still the generated icons at least locally as that's where they linked to by lerna.

jackw replied:
Let's circle back round to this at a later date. I'd rather unblock you so you can iterate on this package.

jackw reviewed (commented):
Awesome work on this @Clarity-89

Clarity-89 commented on packages/grafana-icons/README.md:
```diff
@@ -0,0 +1 @@
+# Grafana Icons
```
Yeah, I was planning to do that once we finalise the package.

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	yarn.lock

Clarity-89 committed:
Add missing deps

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	packages/grafana-ui/package.json
#	yarn.lock

Clarity-89 committed:
Update lockfile

Clarity-89 committed:
Merge branch 'main' into add-icons-package

joshhunt commented on packages/grafana-ui/package.json:
```diff
@@ -53,6 +53,7 @@
     "@grafana/data": "11.0.0-pre",
     "@grafana/e2e-selectors": "11.0.0-pre",
     "@grafana/faro-web-sdk": "^1.3.6",
+    "@grafana/icons": "0.0.0",
```
we probably want this to be `workspace:*` FYI

joshhunt replied:
huh! interesting okay... good catch. worth investigating...

Clarity-89 committed:
Merge branch 'main' into add-icons-package

Clarity-89 committed:
Remove CI actions

Clarity-89 committed:
Remove generated components from VC

Clarity-89 committed:
Fix path

Clarity-89 committed:
Build ESM only

Clarity-89 committed:
Include cjs

Clarity-89 commented:
Made some updates to the PR:
- Generated icons are removed from VC to keep the logic simpler
- Added a digram to illustrate the current flow
- Removed the CI parts to make the scope smaller

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	yarn.lock

leeoniya commented:
hopefully we're retaining most of the bundle size benefits originally gained by pulling the icons out of react components in the first place ;) https://github.com/grafana/grafana/pull/32387.

i assume this allows us to tree shake the ones that are unused?

another possible benefit of having them in plain svg and external is leveraging native browser caching that may be applied to directly loading image assets. also reduces React overhead possibly. are the generated files wrapping the components in `React.memo`, by chance?

Clarity-89 commented:
> i assume this allows us to tree shake the ones that are unused?

Yeah, that's the main goal here, along with making it easier to upload new icons. There's a [design doc](https://docs.google.com/document/d/1nCtUw40DDwZNZwKvSEH3Bo4FkBFFsMgdTIzFvX48-CU) with more details.

>  are the generated icon files wrapping the components in `React.memo`, by chance?

No, they're not, but that could be added.

leeoniya commented:
:+1: 

the previous optimization was working around maintaining the by-string loading of icons. if that constraint is removed, there are definitely better and much more simple solutions :)

Clarity-89 commented:
> 👍
> 
> the previous optimization was working with maintaining the by-string loading of arbitrary icons. if that constraint is removed, there are definitely better and much more straightforward solutions :)

Yep, but good call about possible bundle size/performance implications. We'll need to check that more closely when the package is in a publishable state.

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	yarn.lock

Clarity-89 committed:
Update lockfile

Clarity-89 committed:
Move generated icons into src

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	yarn.lock

Clarity-89 committed:
Update lockfile

joshhunt commented on packages/grafana-icons/package.json:
```diff
@@ -0,0 +1,64 @@
+    "bundle": "rollup -c rollup.config.ts"
+  },
+  "devDependencies": {
+    "@grafana/tsconfig": "^1.3.0-rc1",
+    "@rollup/plugin-node-resolve": "^15.2.3",
+    "@rollup/plugin-typescript": "^11.1.6",
+    "@svgr/babel-plugin-remove-jsx-attribute": "^8.0.0",
+    "@svgr/cli": "^8.1.0",
+    "@svgr/plugin-jsx": "^8.1.0",
+    "@svgr/plugin-prettier": "^8.1.0",
+    "@svgr/plugin-svgo": "^8.1.0",
+    "@types/node": "^20.11.10",
+    "@types/react": "^18.2.48",
+    "@types/react-dom": "^18",
+    "esbuild": "^0.20.0",
+    "prettier": "^3.2.4",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+    "rimraf": "^5.0.5",
+    "rollup": "2.79.1",
+    "rollup-plugin-dts": "^6.1.0",
+    "rollup-plugin-esbuild": "5.0.0",
+    "rollup-plugin-node-externals": "5.0.0",
+    "ts-node": "^10.9.2",
+    "typescript": "^5.3.3"
```
Can you please check that that version of these dependencies match the root workspace package.json? I think especially the esbuild and typescript versions should match the same.

Clarity-89 committed:
Sync deps with grafana

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	yarn.lock

Clarity-89 committed:
Update lockfile

Clarity-89 committed:
Merge branch 'main' into add-icons-package

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	yarn.lock

Clarity-89 committed:
Update lockfile

Clarity-89 committed:
Update package name

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	yarn.lock

Clarity-89 committed:
Cleanup

Clarity-89 committed:
Memoize the components

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	yarn.lock

Clarity-89 committed:
Publish esm only

Clarity-89 commented on packages/grafana-icons/src/IconBase.tsx:
```diff
@@ -0,0 +1,52 @@
+        lineHeight: 0,
+        verticalAlign: 'middle',
+      }}
+      {...props}
```
Good catch!

Clarity-89 committed:
Do not overwrite props

Clarity-89 commented on packages/grafana-icons/package.json:
```diff
@@ -0,0 +1,65 @@
+    "./LICENSE_APACHE2"
+  ],
+  "scripts": {
+    "postinstall": "yarn generate",
```
Just tested and it doesn't run if the install step fails. Since this also generates the index file, we need to make sure it's available as early as possible, otherwise the build will fail. Not sure of there's a better stage to generate icons?

joshhunt replied:
@jackw Could we do this with nx?

Clarity-89 replied:
It seems that excluding this package from workspaces either in package.json or lerna.json, also excludes it from the Nx project list, which makes it impossible to reference this project in the `dependsOn` scripts :(

Clarity-89 committed:
Add generate script to build

Clarity-89 committed:
Update comment

Clarity-89 committed:
Generate typedefs from index.ts

Clarity-89 committed:
test

Clarity-89 committed:
test

Clarity-89 commented on package.json:
```diff
@@ -431,6 +431,7 @@
   "workspaces": {
     "packages": [
       "packages/*",
+      "packages/!(grafana-icons)/**",
```
Can't find the link where I saw this initially, but it was in one of the github issues.
> which isn't really true as yarn continues to symlink the package yet Lerna should ignore it for versioning and publishing?

This is exactly how this works and it looks like a difference in how the same config option is read by lerna and yarn workspaces. I agree this is confusing, maybe moving this to lerna.json is better as it clearly signals that this option is lerna-specific, but then there might be a confusion that one of the workspaces mentioned in package.json is not being versioned?

jackw replied:
Personally I'd move it to the lerna.json to give clear separation of what is yarns config and what is lernas. But we can do that later once we've ironed out a few more kinks.

Clarity-89 committed:
Restore

Clarity-89 commented on packages/grafana-icons/.svgrrc.js:
```diff
@@ -0,0 +1,32 @@
+          '@svgr/babel-plugin-remove-jsx-attribute',
+          {
+            elements: ['path', 'g', 'clipPath'],
+            attributes: ['id', 'fill'],
```
That seems to break a few of the custom icons, like gf-traces, so better keep it for now.

jackw replied:
Fair enough, just wanted to highlight it. 👍

Clarity-89 committed:
Add more complex icons

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	yarn.lock

Clarity-89 committed:
Add readme

joshhunt reviewed (approved):
I don't think `preinstall` is the best place to do the generation, but I think we can continue to iterate on it.

I'm curious to see if we can utilise nx to build it as a dependency.

I would also like an 👍 from Jack before having it merged in.

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	package.json
#	yarn.lock

Clarity-89 committed:
Update lockfile

jackw commented on packages/grafana-icons/package.json:
```diff
@@ -0,0 +1,65 @@
+    "url": "https://github.com/grafana/grafana.git",
+    "directory": "packages/grafana-icons"
+  },
+  "main": "src/index.ts",
```
To make this an ESM only package I believe we need to change a few things. We'd need to add `"type": "module"` then either keep main or replace it with `exports`. We wouldn't need publishConfig.module if we only build for ESM. Maybe we do this as a follow up.

jackw commented on packages/grafana-icons/.svgrrc.js:
```diff
@@ -0,0 +1,32 @@
+        [
+          '@svgr/babel-plugin-remove-jsx-attribute',
+          {
+            elements: ['path', 'g', 'clipPath'],
```
Fair enough, just thought it worth highlighting.

jackw commented on packages/grafana-icons/package.json:
```diff
@@ -0,0 +1,64 @@
+{
+  "name": "@grafana/saga-icons",
+  "version": "0.0.0",
```
I think the best way to iterate on this package is to make it private right now and get this merged in:

```suggestion
  "version": "0.0.0",
  "private": true,
```

jackw reviewed (approved):
I'm going to approve it but before merge I would like that:

1. we make this package private so we can figure out a few of the remaining build questions without worry a bump to yarn or lerna will attempt to publish / break releases somewhere.
2. sort out the postInstall script - probably by using nx instead.

Clarity-89 committed:
Make the package private

Clarity-89 committed:
Move package exclusion to lerna.json

Clarity-89 committed:
Remove exclusion

Clarity-89 committed:
Revert

Clarity-89 committed:
test packages

Clarity-89 committed:
Fix

Clarity-89 committed:
Fix

Clarity-89 committed:
Revert

Clarity-89 committed:
Move generate icons script to the root

Clarity-89 committed:
Merge branch 'main' into add-icons-package

Clarity-89 committed:
Update script

Clarity-89 committed:
Merge branch 'main' into add-icons-package

Clarity-89 committed:
Merge branch 'main' into add-icons-package

# Conflicts:
#	yarn.lock

Clarity-89 committed:
Remove icons script

Clarity-89 committed:
Merge branch 'main' into add-icons-package

Clarity-89 committed:
Move icons exclusion to package.json
